### PR TITLE
feat: Add database fixtures for testing migrations

### PR DESCRIPTION
**Pull request by mafredri**
This PR adds two types of fixtures for testing database migrations

1. Incremental fixtures in `coderd/database/migrations/testdata/fixtures`
2. Full deployment fixtures in `coderd/database/migrations/testdata/full_dumps/vX.X.X`

For 1), the purpose of the `fixtures` folder is for us to incrementally add data we feel should be tested during a migration, it's based on a complete dump, but the idea is to evolve it via manual additions (see 000024 example).

For 2), the purpose is for us to easily take a specific version of Coder, create a deployment, dump the database, and ensure that this deployment upgrades without issue for all eternity.

### TODO

- [x] Documentation
    - [x] Describe the logic behind `000022` in file names and how to add fixtures
    - [x] Describe how to create full dumps
- [x] (In this or future PR) Checks to ensure new tables are populated with data?

### Questions

Thoughts on where to place the documentation?

- https://coder.com/docs/coder-oss/latest/CONTRIBUTING ?
- A README.md in `coderd/datdabase/migrations`?

### CODE REVIEW

**Review by kylecarbs — APPROVED** [review]
I enjoy the automated nature of this - it'll just tell us when we broke prior versions!

How do you imagine we'll promote devs to add a new dump or fixture?

**Comment by mafredri** [comment]
> How do you imagine we'll promote devs to add a new dump or fixture?

How about something simple like checking table stats at the end of the test?

```
coder=# select relname, n_tup_ins from pg_stat_user_tables where n_tup_ins = 0;
  relname   | n_tup_ins
------------+-----------
 licenses   |         0
 audit_logs |         0
```

We can start off with a allow-list for empty tables, but when new ones are added, if there are no rows at the end of migration + fixture, return an error?

It might be a bit annoying for this to be a failing test though, as opposed to a linting event. But we could maybe remedy that by running something like `go test ./coderd/database/migrations -run=...Fixtures` before the main tests (or as part of lint).

Regarding dumps, not sure. I think there needs to be an intent of "I want this type of Coder deployment to be tested" for a full dump to make sense.

We could gather table stats from both fixtures and full dumps, in this case when you run into a "lint" issue for a table, you'd be presented with the option of creating a full dump or a simple fixture.

**Comment by mafredri** [comment]
@kylecarbs with 7d1c3e8 it looks like this:

```
❯ go test ./coderd/database/migrations -run=TestMigrateUpWithFixtures
--- FAIL: TestMigrateUpWithFixtures (0.00s)
    migrate_test.go:257: The following tables have zero rows, consider adding fixtures for them or create a full database dump:
    migrate_test.go:258: tables have zero rows: [audit_logs git_auth_links group_members licenses replicas]
    migrate_test.go:260: See https://github.com/coder/coder/blob/main/migrations/README.md#creating-fixtures for more information
FAIL
FAIL	github.com/coder/coder/coderd/database/migrations	7.608s
FAIL
```

File: coderd/database/migrations/migrate_test.go [hunk 1]
```diff
@@ -129,3 +139,197 @@ func TestCheckLatestVersion(t *testing.T) {
...
+				WHERE
+					schemaname != 'information_schema'
+					AND schemaname != 'pg_catalog'
+					AND tablename NOT LIKE 'test_migrate_%'
+			`).Scan(&tables)
+			require.NoError(t, err)
+
+			for _, table := range tables {
+				var count int
+				err = db.QueryRowContext(ctx, "SELECT COUNT(*) FROM "+table).Scan(&count)
```
Comment (mafredri) [inline coderd/database/migrations/migrate_test.go:326]: A slightly uglier method than checking pg stats, problem with stats is we don't control when they're updated.

**Comment by mafredri** [comment]
This PR is now pretty much complete:
- We have docs for adding fixtures
- We detect if new tables are added without fixtures and return a descriptive error

The only remaining issue would be to improve how fast a user might run into the migration/fixture error, but I think we can improve that when/if it becomes annoying.

@bpmct would love your input on this (esp. the docs part).

**Pull request merged**